src/gui/owncloudgui.cpp

```diff
@@ -400,41 +400,159 @@ void ownCloudGui::addAccountContextMenu(AccountStatePtr accountState, QMenu *men
-    return !var.isEmpty();
-}
-
-
 void ownCloudGui::slotContextMenuAboutToShow()
 {
     // For some reason on OS X _contextMenu->isVisible returns always false
     qDebug() << "";
-    _contextMenuVisible = true;
+    _contextMenuVisibleOsx = true;
 }
 
 void ownCloudGui::slotContextMenuAboutToHide()
 {
     // For some reason on OS X _contextMenu->isVisible returns always false
     qDebug() << "";
-    _contextMenuVisible = false;
+    _contextMenuVisibleOsx = false;
+}
+
+bool ownCloudGui::contextMenuVisible() const
+{
+#ifdef Q_OS_MAC
+    return _contextMenuVisibleOsx;
+#else
+    return _contextMenu->isVisible();
+#endif
+}
+
+static bool minimalTrayMenu()
+{
+    static QByteArray var = qgetenv("OWNCLOUD_MINIMAL_TRAY_MENU");
+    return !var.isEmpty();
+}
+
+static bool updateWhileVisible()
+{
+    static QByteArray var = qgetenv("OWNCLOUD_TRAY_UPDATE_WHILE_VISIBLE");
+    if (var == "1") {
+        return true;
+    } else if (var == "0") {
+        return false;
+    } else {
+#ifdef Q_OS_MAC
+        return false; // https://bugreports.qt.io/browse/QTBUG-54845
+#else
+        return true;
+#endif
+    }
+}
+
+static QByteArray forceQDBusTrayWorkaround()
+{
+    static QByteArray var = qgetenv("OWNCLOUD_FORCE_QDBUS_TRAY_WORKAROUND");
+    return var;
 }
 
 void ownCloudGui::setupContextMenu()
 {
+    if (_contextMenu) {
+        return;
+    }
+
+    _contextMenu.reset(new QMenu());
+    _contextMenu->setTitle(Theme::instance()->appNameGUI() );
+
+    _recentActionsMenu = new QMenu(tr("Recent Changes"), _contextMenu.data());
+
+    // this must be called only once after creating the context menu, or
+    // it will trigger a bug in Ubuntu's SNI bridge patch (11.10, 12.04).
+    _tray->setContextMenu(_contextMenu.data());
+
     // The tray menu is surprisingly problematic. Being able to switch to
     // a minimal version of it is a useful workaround and testing tool.
     if (minimalTrayMenu()) {
-        if (!_contextMenu) {
-            _contextMenu.reset(new QMenu());
-            _recentActionsMenu = new QMenu(tr("Recent Changes"), _contextMenu.data());
-            _tray->setContextMenu(_contextMenu.data());
-            _contextMenu->addAction(_actionQuit);
+        _contextMenu->addAction(_actionQuit);
+        return;
+    }
+
+    // Enables workarounds for bugs introduced in Qt 5.5.0
+    // In particular QTBUG-47863 #3672 (tray menu fails to update and
+    // becomes unresponsive) and QTBUG-48068 #3722 (click signal is
+    // emitted several times)
+    // The Qt version check intentionally uses 5.0.0 (where platformMenu()
+    // was introduced) instead of 5.5.0 to avoid issues where the Qt
+    // version used to build is different from the one used at runtime.
+    // If we build with 5.6.1 or newer, we can skip this because the
+    // bugs should be fixed there.
+#ifdef Q_OS_LINUX
+#if (QT_VERSION >= QT_VERSION_CHECK(5, 0, 0)) && (QT_VERSION < QT_VERSION_CHECK(5, 5, 1))
```

Comment:
Technicly, one can compile an app against 5.5.1 and run it against 5.5.0.  so i guess this should be < 5.6